extensions/src/main/java/org/pentaho/platform/plugin/services/email/EmailService.java

```diff
@@ -186,18 +237,37 @@ protected PasswordAuthentication getPasswordAuthentication() {
       msg.setText( messages.getString( "EmailService.MESSAGE" ) );
       msg.setHeader( "X-Mailer", "smtpsend" );
       msg.setSentDate( new Date() );
-      Transport.send( msg );
-      sendEmailMessage = "EmailTester.SUCESS";
+      sendEmail( emailConfig, token, session, msg );
+      sendEmailMessage = TEST_EMAIL_SUCCESS;
     } catch ( Exception e ) {
       logger.error( messages.getString( "EmailService.NOT_CONFIGURED" ), e );
-      sendEmailMessage = "EmailTester.FAIL";
+      sendEmailMessage = TEST_EMAIL_FAIL;
     }
+    setEmailConfig( emailConfig );
     return sendEmailMessage;
   }
 
+  public void sendEmail( final IEmailConfiguration emailConfig, IEmailAuthenticationResponse token, Session session, MimeMessage msg ) throws Exception {
```

Comment:
I feel like we should use OOP here instead of having bunch of if/else statements. I'm sure that was the original author's intention of creating IEmailService.java and IEmailConfiguration.java was. Imagine when we add another email configuration these functions (sendEmail() and sendEmailTest()) and  will get complex and hard to troubleshoot and hard to test.

With this current approach, every new email protocol and email config properties needs to update the IEmailService.java interface which ruins the benefits of OOP and creates more work to keep updating.

I would have a class 
SMTPService.java extends IEmailService that implements sendEmail(IEmailConfiguration emailConfig)
handles both cases for "SMTP" and "SMTPS" and MicrosoftGraphApiService.java that extends IEmailService that implements sendEmail(IEmailConfiguration emailConfig).

Maybe keep EmailService as it, since there probably is some shared logic that most classes can share.

Reply:
Agree with your interpretation of the likely original intent. The current/original interfaces seem independent of specific authentication mechanisms.

---

Agree with multiple implementations of a generic `IEmailService`.
Today, when obtaining an instance of `IEmailService` from the PentahoSystem, the registration with the _key_ (bean id) `"IEmailService"` is requested:

```java
IEmailService service = PentahoSystem.get( IEmailService.class, "IEmailService", session );
```

In the new approach, one would request for the registration having a certain property set. For example:
```java
PentahoSystem.get( IEmailService.class, session, Collections.singletonMap( "authType", "XOAUTH2" ) );
```

For backward compatibility, we could fallback to the old method, but only for the legacy authentication types, failing for the new authentication types.

---

What about `IEmailConfiguration`? Should this still become a single "mashup" interface, including configuration for all possible auth. mechanisms? Or, should sub-interfaces be defined for specific sub-types of configurations? Specific service implementations would require corresponding sub-types of configuration?

Reply:
Regarding the compatibility issue, the original `IEmailService` was only used for sending test emails. Sending emails proper was being done in the `Emailer` class. The latter, though, is not pluggable and so customers had no way to change the implementation of how emails are actually sent. The only other functionality offered by `IEmailService` is persisting and loading the email configuration, and it is not seen as something of special interest for customers.

Concluding, it is very unlikely that any customer implemented `IEmailService`, and so this is not considered to be a real breaking change.